util/fs.go

```diff
@@ -0,0 +1,226 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package util
+
+import (
+	"archive/zip"
+	"fmt"
+	"io"
+	"io/fs"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+// PackageFile is the interface that files in the file system need to implement.
+// Seeker interface is needed for http helpers to serve files.
+type PackageFile = io.ReadSeekCloser
+
+type PackageFileSystem interface {
+	Stat(name string) (os.FileInfo, error)
+	Open(name string) (PackageFile, error)
+	Glob(pattern string) (matches []string, err error)
+	Close() error
+}
+
+func NewPackageFileSystem(path string) (PackageFileSystem, error) {
+	if path == "" {
+		return NewVirtualPackageFileSystem()
+	}
+	info, err := os.Stat(path)
+	if err != nil {
+		return nil, err
+	}
+	switch {
+	case info.IsDir():
+		return NewExtractedPackageFileSystem(path)
+	case strings.HasSuffix(path, ".zip"):
+		return NewZipPackageFileSystem(path)
```

Comment:
That reminded me: https://github.com/golang/tools/blob/master/godoc/vfs/zipfs/zipfs.go

Reply:
Oh, I was hoping to find something like this :slightly_smiling_face: With the new interfaces added in 1.16 these things should be more common.